File: apps/vm_runner/build_job.ts
import {
    DocumentSnapshot,
    Firestore,
    QuerySnapshot,
} from "npm:firebase-admin/firestore";
import { baseUrl } from "./prod_urls.ts";

export async function getBuildJob(db: Firestore): Promise<QuerySnapshot> {
    return await db
        .collection("build_jobs")
        .where("buildStatus", "==", "queued")
        .orderBy("createdAt", "asc")
        .limit(1)
        .get();
}

export async function getWorkflowDocs(
    db: Firestore,
    workflowId: string,
): Promise<DocumentSnapshot> {
    return await db.collection("workflows").doc(workflowId).get();
}

export async function setStatusToInProgress(
    db: Firestore,
    jobId: string,
): Promise<void> {
    await updateBuildStatus(jobId, "inProgress");
    await db.collection("build_jobs").doc(jobId).update({
        buildStatus: "inProgress",
    });
}

export async function setStatusToSuccess(
    db: Firestore,
    jobId: string,
): Promise<void> {
    await updateBuildStatus(jobId, "success");
    await db.collection("build_jobs").doc(jobId).update({
        buildStatus: "success",
    });
}

export async function setStatusToFailure(
    db: Firestore,
    jobId: string,
): Promise<void> {
    await updateBuildStatus(jobId, "failure");
    await db.collection("build_jobs").doc(jobId).update({
        buildStatus: "failure",
    });
}

export async function updateBuildStatus(
    jobId: string,
    status: string,
): Promise<void> {
    const url = new URL("/update_checks", baseUrl);

    const body = JSON.stringify({
        jobId: jobId,
        checksStatus: status,
    });

    try {
        const response = await fetch(url, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: body,
        });

        if (!response.ok) {
            throw new Error(
                `APIエラー: ${response.status} ${response.statusText}`,
            );
        }

        console.log(`Build status updated: ${status}`);
    } catch (error) {
        console.error(`Failed to update build status: ${error}`);
    }
}

File: apps/vm-runner/build_job.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	"cloud.google.com/go/firestore"
	"github.com/getsentry/sentry-go"
)

type BuildJob struct {
	ID     string `json:"id"`
	GitHub struct {
		RepoURL        string `json:"repositoryUrl"`
		InstallationId int    `json:"installationId"`
		AppId          int    `json:"appId"`
		BuildBranch    string `json:"buildBranch"`
	} `json:"github"`
	WorkflowId string `json:"workflowId"`
	Status     string `json:"status"`
}

func ParseBuildJob(jsonData []byte) (*BuildJob, error) {
	var job BuildJob

	err := json.Unmarshal(jsonData, &job)
	if err != nil {
		return nil, fmt.Errorf("failed to unmarshal build job: %w", err)
	}

	fmt.Printf("Successfully parsed BuildJob: %v\n", job)

	return &job, nil
}

func GetBuildJob(ctx context.Context, firestoreClient *firestore.Client, infoLogger *log.Logger) (*BuildJob, error) {
	var snap *firestore.DocumentSnapshot

	err := firestoreClient.RunTransaction(ctx, func(ctx context.Context, tx *firestore.Transaction) error {
		ref := firestoreClient.Collection("build_jobs_v3").
			Where("buildStatus", "==", "queued").
			OrderBy("createdAt", firestore.Asc).
			Limit(1)

		docs, err := ref.Documents(ctx).GetAll()
		if err != nil {
			return fmt.Errorf("failed to fetch build jobs: %v", err)
		}

		if len(docs) == 0 {
			return fmt.Errorf("no queued build jobs found")
		}

		docRef := docs[0].Ref
		freshDocSnap, err := tx.Get(docRef)
		if err != nil {
			return fmt.Errorf("failed to get document snapshot: %v", err)
		}

		if !freshDocSnap.Exists() {
			return fmt.Errorf("document does not exist")
		}

		data := freshDocSnap.Data()
		freshStatus, ok := data["buildStatus"].(string)
		if !ok {
			return fmt.Errorf("invalid buildStatus field")
		}

		if freshStatus == "queued" {
			err = tx.Update(docRef, []firestore.Update{
				{
					Path:  "buildStatus",
					Value: "inProgress",
				},
				{
					Path:  "updatedAt",
					Value: time.Now(),
				},
			})
			if err != nil {
				return fmt.Errorf("failed to update document: %v", err)
			}
			infoLogger.Printf("Document %s updated to inProgress.", docRef.ID)
		} else {
			return fmt.Errorf("build job status is not queued: %s", freshStatus)
		}

		snap = freshDocSnap
		return nil
	})

	if err != nil {
		return nil, err
	}

	if snap == nil {
		return nil, fmt.Errorf("no valid build job found")
	}

	data := snap.Data()

	jsonData, err := json.Marshal(data)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal workflow data: %v", err)
	}

	job, err := ParseBuildJob(jsonData)
	if err == nil {
		infoLogger.Printf("Successfully parsed BuildJob: %v\n", job)
	} else {
		sentry.CaptureMessage(fmt.Sprintf("Failed to parse BuildJob: %v", err))
		return nil, fmt.Errorf("failed to parse BuildJob: %v", err)
	}

	return job, nil
}
